lib/job/arUpdatePublicationStatusJob.class.php

```diff
@@ -18,42 +18,113 @@
+ * Updates the publication status to the descendants of an information object. Uses
+ * ES updateByQuery to update the ES index rather than replace each child document.
  */
 class arUpdatePublicationStatusJob extends arBaseJob
 {
     /**
      * @see arBaseJob::$requiredParameters
      */
-    protected $extraRequiredParameters = ['objectId', 'publicationStatusId'];
+    protected $extraRequiredParameters = [
+        'objectId',
+        'publicationStatusId',
+    ];
 
     public function runJob($parameters)
     {
-        if (null === $resource = QubitInformationObject::getById($parameters['objectId'])) {
-            $this->error($this->i18n->__('Invalid description id: %1', ['%1' => $parameters['objectId']]));
+        if (null === $publicationStatus = QubitTerm::getById($parameters['publicationStatusId'])) {
+            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
 
             return false;
         }
 
-        if (null === $publicationStatus = QubitTerm::getById($parameters['publicationStatusId'])) {
-            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
+        if (null === $resource = QubitInformationObject::getById($parameters['objectId'])) {
+            $this->error($this->i18n->__('Invalid description id: %1', ['%1' => $parameters['objectId']]));
 
             return false;
         }
 
-        $message = $this->i18n->__('Updating publication status for the descendants of "%1" to "%2".', ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]);
+        $sql = 'UPDATE status
+            JOIN information_object io ON status.object_id = io.id
+            SET status.status_id = :publicationStatus
+            WHERE status.type_id = :publicationStatusType
+            AND io.lft > :lft
+            AND io.rgt < :rgt';
+
+        $message = $this->i18n->__(
+            'Updating publication status for the descendants of "%1" to "%2".',
+            ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]
+        );
+
         $this->job->addNoteText($message);
         $this->info($message);
 
-        $descriptionsUpdated = 0;
-        foreach ($resource->descendants as $descendant) {
-            $descendant->setPublicationStatus($publicationStatus->id);
-            $descendant->save();
+        $params = [
+            ':publicationStatus' => $publicationStatus->id,
+            ':publicationStatusType' => QubitTerm::STATUS_TYPE_PUBLICATION_ID,
+            ':lft' => $resource->lft,
+            ':rgt' => $resource->rgt,
+        ];
+
+        $descriptionsUpdated = QubitPdo::modify(
+            $sql,
+            $params
+        );
+
+        $response = QubitSearch::getInstance()->updateByQuery(
+            [
+                'ancestors' => $resource->id,
+            ],
```

Comment:
Nitpick: no need to split this array into multiple lines.

Reply:
This has been removed.